drivers/i2c/i2c_sam_twi.c

```diff
@@ -0,0 +1,390 @@
+
+	/* Enable Master Mode */
+	twi->TWI_CR = TWI_CR_MSEN;
+
+	return 0;
+}
+
+static void write_msg_start(Twi *const twi, struct twi_msg *msg, u8_t daddr)
+{
+	/* Set slave address and number of internal address bytes. */
+	twi->TWI_MMR = TWI_MMR_DADR(daddr);
+
+	/* Write first data byte on I2C bus */
+	twi->TWI_THR = msg->buf[msg->idx++];
+
+	/* Enable Transmit Ready and Transmission Completed interrupts */
+	twi->TWI_IER = TWI_IER_TXRDY | TWI_IER_TXCOMP | TWI_IER_NACK;
+}
+
+static void read_msg_start(Twi *const twi, struct twi_msg *msg, u8_t daddr)
+{
+	u32_t twi_cr_stop;
+
+	/* Set slave address and number of internal address bytes */
+	twi->TWI_MMR = TWI_MMR_MREAD | TWI_MMR_DADR(daddr);
+
+	/* In single data byte read the START and STOP must both be set */
+	twi_cr_stop = (msg->len == 1) ? TWI_CR_STOP : 0;
+	/* Start the transfer by sending START condition */
+	twi->TWI_CR = TWI_CR_START | twi_cr_stop;
+
+	/* Enable Receive Ready and Transmission Completed interrupts */
+	twi->TWI_IER = TWI_IER_RXRDY | TWI_IER_TXCOMP | TWI_IER_NACK;
+}
+
+static int i2c_sam_twi_transfer(struct device *dev, struct i2c_msg *msgs,
+				u8_t num_msgs, u16_t addr)
+{
+	const struct i2c_sam_twi_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct i2c_sam_twi_dev_data *const dev_data = DEV_DATA(dev);
+	Twi *const twi = dev_cfg->regs;
+
+	__ASSERT_NO_MSG(msgs);
+	if (!num_msgs) {
+		return 0;
+	}
+
+	/* Clear pending interrupts, such as NACK. */
+	(void)twi->TWI_SR;
+
+	/* Set number of internal address bytes to 0, not used. */
+	twi->TWI_IADR = 0;
+
+	for (; num_msgs > 0; num_msgs--, msgs++) {
+		dev_data->msg.buf = msgs->buf;
+		dev_data->msg.len = msgs->len;
+		dev_data->msg.idx = 0;
+		dev_data->msg.twi_sr = 0;
+		dev_data->msg.flags = msgs->flags;
+
+		/*
+		 * REMARK: Dirty workaround:
+		 *
+		 * The controller does not have a documented, generic way to
+		 * issue RESTART when changing transfer direction as master.
+		 * Send a stop condition in such a case.
+		 */
+		if (num_msgs > 1) {
+			if ((msgs[0].flags & I2C_MSG_RW_MASK) !=
+			    (msgs[1].flags & I2C_MSG_RW_MASK)) {
+				dev_data->msg.flags |= I2C_MSG_STOP;
+			}
+		}
+
+		if ((msgs->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+			read_msg_start(twi, &dev_data->msg, addr);
+		} else {
+			write_msg_start(twi, &dev_data->msg, addr);
+		}
+		/* Wait for the transfer to complete */
+		k_sem_take(&dev_data->sem, K_FOREVER);
+
+		if (dev_data->msg.twi_sr > 0) {
+			/* Something went wrong */
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+static void i2c_sam_twi_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct i2c_sam_twi_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	struct i2c_sam_twi_dev_data *const dev_data = DEV_DATA(dev);
+	Twi *const twi = dev_cfg->regs;
+	struct twi_msg *msg = &dev_data->msg;
+	u32_t isr_status;
+
+	/* Retrieve interrupt status */
+	isr_status = twi->TWI_SR & twi->TWI_IMR;
+
+	/* Not Acknowledged */
+	if (isr_status & TWI_SR_NACK) {
+		msg->twi_sr = isr_status;
```

Comment:
Do you need to give the semaphore here?

Reply:
According to the datasheet TXCOMP is always set alongside NACK but the code indeed could have been written better. I updated it to make it more explicit.